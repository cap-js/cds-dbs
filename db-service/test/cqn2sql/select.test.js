'use strict'
const cds = require('@sap/cds/lib')
const cqn2sql = require('../../lib/cqn2sql')
const cqn = require('./cqn.js')

// const getExpected = (sql, values) => {
//   return {
//     sql: sql,
//     values: values || []
//   }
// }

describe('cqn2sql', () => {
  beforeAll(async () => {
    cds.model = await cds.load(__dirname + '/testModel').then(cds.linked)
  })
  describe('selection of columns of one table', () => {
    test('with select with from ref and elements = undefined', () => {
      const { sql } = cqn2sql(cqn.select)
      expect(sql).toMatchSnapshot()
    })

    // empty columns will be ignored
    test('with select with from ref and elements as empty array', () => {
      const cqnSelect = {
        SELECT: Object.assign({}, cqn.select.SELECT, { columns: [] }),
      }
      const { sql } = cqn2sql(cqnSelect)
      expect(sql).toMatchSnapshot()
    })

    test('with select specific elements with from ref', () => {
      const { sql } = cqn2sql(cqn.selectWithColumns)
      expect(sql).toMatchSnapshot()
    })

    // empty orderBy will be ignored
    test('with select with empty orderBy and specific elements with from type string', () => {
      const { sql } = cqn2sql(cqn.selectWithColumnsEmptyOrderBy)
      expect(sql).toMatchSnapshot()
    })

    test('with select with asterisk in columns', () => {
      const { sql } = cqn2sql(cqn.selectWithColumnsWithAsterisk)
      expect(sql).toMatchSnapshot()
    })

    test('select distinct', () => {
      const { sql } = cqn2sql(cqn.selectDistinct)
      expect(sql).toMatchSnapshot()
    })

    test('with select that has (faked) reflection model', () => {
      const { sql } = cqn2sql(cqn.selectWithCSN)
      expect(sql).toMatchSnapshot()
    })

    test('with select from non existent entity with star wildcard', () => {
      expect(() => {
        let q = cqn.selectNonExistent
        // Skip cqn4sql as infer requires the entity to exist
        const render = q => new cqn2sql.class().render(q)
        const { sql } = render(q)
        expect(sql).toMatchSnapshot()
        q = cds.ql.clone(q)
        q.SELECT.expand = 'root'
        render(q) // throws
      }).toThrowError('Query was not inferred and includes expand. For which the metadata is missing.')
    })

    test('with select from non existent entity with star wildcard (extended)', () => {
      expect(() => {
        const customCqn2sql = class extends cqn2sql.class {
          SELECT_columns({ SELECT }) {
            return SELECT.columns.map(x => `${this.quote(this.column_name(x))}`)
          }
        }
        const q = cqn.selectNonExistent
        // Skip cqn4sql as infer requires the entity to exist
        const render = q => new customCqn2sql().render(q)
        render(q) // throws
      }).toThrowError(
        `Query was not inferred and includes '*' in the columns. For which there is no column name available.`,
      )
    })
  })

  describe('WHERE', () => {
    test('entries where one column holds entries smaller than 9', () => {
      const { sql } = cqn2sql(cqn.selectWhereSmaller)
      expect(sql).toMatchSnapshot()
    })

    test('entries where with place holder', () => {
      const { sql, values } = cqn2sql(cqn.selectWhereWithOnePlaceholderCqn)
      expect({ sql, values }).toMatchSnapshot()
    })

    test('entries where with int reference and param true', () => {
      const { sql, values } = cqn2sql(cqn.selectWhereWithRefIntParamTrueCqn)
      expect({ sql, values }).toMatchSnapshot()
    })

    test('where with partial cqn', () => {
      const { sql } = cqn2sql(cqn.selectWhereCqn)
      expect(sql).toMatchSnapshot()
    })

    test('where with two partial cqn', () => {
      const { sql } = cqn2sql(cqn.selectWhereTwoCqn)
      expect(sql).toMatchSnapshot()
    })

    test('entries where one column holds entries which are in list', () => {
      const { sql } = cqn2sql(cqn.selectWhereList)
      expect(sql).toMatchSnapshot()
    })

    test('select with a nested select in where', () => {
      const { sql } = cqn2sql(cqn.selectWhereSelect)
      expect(sql).toMatchSnapshot()
    })

    test('select with a nested select in a complex where', () => {
      const { sql } = cqn2sql(cqn.selectComplexWhere)
      expect(sql).toMatchSnapshot()
    })

    test('with select with exist in where condition', () => {
      const { sql } = cqn2sql(cqn.selectWhereExists)
      expect(sql).toMatchSnapshot()
    })

    test('with list of values', () => {
      const cqn = {
        SELECT: {
          from: { ref: ['Foo'] },
          where: [
            { list: [{ ref: ['a'] }, { ref: ['b'] }, { val: 1 }] },
            '=',
            { list: [{ ref: ['c'] }, { val: 'd' }, { ref: ['x'] }] },
          ],
        },
      }
      const { sql, values } = cqn2sql(cqn)
      expect({ sql, values }).toMatchSnapshot()
    })

    test('with contains with one column in where clause', () => {
      // const expected = getExpected("SELECT Foo.ID,Foo.a,Foo.b,Foo.c,Foo.x FROM Foo AS Foo FROM T WHERE ( b LIKE ( '%' || ? || '%' ) ESCAPE '^' )", ['5'])
      const { sql, values } = cqn2sql(cqn.selectContainsOneColumn)
      expect({ sql, values }).toMatchSnapshot()
    })

    test('EXISTS with nested EXISTS', () => {
      const { sql } = cqn2sql(cqn.selectWhereNestedExists)
      expect(sql).toMatchSnapshot()
    })

    test('with function without alias', () => {
      const toThrow = () => {
        return cqn2sql(cqn.selectFuncitonWithoutAlias)
      }
      expect(toThrow).toThrowError('Expecting expression to have an alias name')
    })

    test('with contains with multiple values', () => {
      const cqn = {
        SELECT: {
          from: { ref: ['Foo'] },
          where: [
            { ref: ['a'] },
            '=',
            { val: 0 },
            'and',
            {
              func: 'contains',
              args: [{ list: [{ ref: ['a'] }] }, { val: 'z' }, 'or', { val: 'zz' }],
            },
          ],
        },
      }
      const toThrow = () => {
        return cqn2sql(cqn)
      }
      expect(toThrow).toThrowError('Unsupported expr: or')
    })

    test('with contains with multiple arguments', () => {
      const cqn = {
        SELECT: {
          from: { ref: ['Foo'] },
          where: [
            { ref: ['a'] },
            '=',
            { val: 0 },
            'and',
            {
              func: 'contains',
              args: [{ list: [{ ref: ['a'] }, { ref: ['b'] }, { ref: ['c'] }, { ref: ['x'] }] }, { val: 'z' }],
            },
          ],
        },
      }
      const { sql } = cqn2sql(cqn)
      expect(sql).toMatchSnapshot()
    })
  })
  describe('HAVING clauses', () => {
    test('with select specific elements with from type string with having clause', () => {
      const { sql } = cqn2sql(cqn.selectHaving)
      expect(sql).toMatchSnapshot()
    })
  })

  describe('complex combinations', () => {
    test('WHERE, GROUP BY, HAVING, ORDER BY, LIMIT, OFFSET', () => {
      const { sql } = cqn2sql(cqn.selectAggregationLimitOrder)
      expect(sql).toMatchSnapshot()
    })

    test('AS, sub query', () => {
      const { sql } = cqn2sql(cqn.selectSubSelect)
      expect(sql).toMatchSnapshot()
    })

    test('Exists in object mode in complex where', () => {
      const { sql, values } = cqn2sql(cqn.selectComplexWhereWithExists)
      expect({ sql, values }).toMatchSnapshot()
    })
  })

  describe('GROUP BY', () => {
    test('GROUP BY two columns', () => {
      const { sql } = cqn2sql(cqn.groupBy)
      expect(sql).toMatchSnapshot()
    })
  })

  describe('ORDER BY', () => {
    test('ORDER BY alias', () => {
      const { sql } = cqn2sql(cqn.orderByWithAlias)
      expect(sql).toMatchSnapshot()
    })
  })

  describe('ONE', () => {
    test('one results in limit 1', () => {
      const { sql } = cqn2sql(cqn.one)
      expect(sql).toMatchSnapshot()
    })

    test('one with additional limit with offset', () => {
      // Original DB layer expectation is to mix limit and one
      // One has priority over limit.rows, but limit.offset is still applied
      const { sql } = cqn2sql(cqn.oneWithLimit)
      expect(sql).toEqual('SELECT Foo.a,Foo.b,Foo.c FROM Foo as Foo LIMIT 1 OFFSET 5')
    })
  })

  describe('LIMIT', () => {
    test('with limit without offset', () => {
      const { sql } = cqn2sql(cqn.limit)
      expect(sql).toMatchSnapshot()
    })

    test('with limit and offset', () => {
      const { sql } = cqn2sql(cqn.limitOffset)
      expect(sql).toMatchSnapshot()
    })

    test('limit without rows throws error', () => {
      const toThrow = () => {
        return cqn2sql({ SELECT: { from: { ref: ['Foo'] }, limit: { offset: { val: 5 } } } })
      }
      expect(toThrow).toThrowError('Rows parameter is missing in SELECT.limit(rows, offset)')
    })
  })

  describe('aggregation functions', () => {
    test('with select with same functions without alias in elements', () => {
      const toThrow = () => {
        return cqn2sql(cqn.selectWithSameFunctionsWithoutAlias)
      }
      expect(toThrow).toThrowError('Duplicate definition of element “count”')
    })

    test('with select with different functions without alias in elements', () => {
      const { sql } = cqn2sql(cqn.selectWithFunctionsWithoutAlias)
      expect(sql).toMatchSnapshot()
    })

    test('with select with functions in elements new notation', () => {
      const { sql } = cqn2sql(cqn.selectWithAggregationNew)
      expect(sql).toMatchSnapshot()
    })

    test('with select with count(1)', () => {
      const { sql } = cqn2sql(cqn.selectWithCountOne)
      expect(sql).toMatchSnapshot()
    })

    test('with select with functions in where clause new notation', () => {
      const { sql } = cqn2sql(cqn.selectWhereAggregationNew)
      expect(sql).toMatchSnapshot()
    })
  })

  describe('functions new notation', () => {
    test('in orderby with 1 arg new notation', () => {
      const { sql } = cqn2sql({
        SELECT: {
          from: { ref: ['Foo'] },
          orderBy: [{ func: 'lower', args: [{ ref: ['c'] }], sort: 'desc' }],
        },
      })
      expect(sql).toMatchSnapshot()
    })

    test('in filter with 1 arg new notation', () => {
      const { sql, values } = cqn2sql({
        SELECT: {
          from: { ref: ['Foo'] },
          where: [{ func: 'lower', args: [{ ref: ['c'] }] }, '=', { val: 'name' }],
        },
      })
      expect({ sql, values }).toMatchSnapshot()
    })

    test('in filter with asterisk as arg new notation', () => {
      const { sql } = cqn2sql({
        SELECT: {
          from: { ref: ['Foo'] },
          having: [{ func: 'count', args: ['*'] }, '>', { val: 1 }],
        },
      })
      expect(sql).toMatchSnapshot()
    })

    test('in filter with 2 arg new notation', () => {
      const { sql } = cqn2sql({
        SELECT: {
          from: { ref: ['Foo'] },
          where: [{ ref: ['c'] }, '=', { func: 'concat', args: [{ ref: ['a'] }, { ref: ['b'] }] }],
        },
      })
      expect(sql).toMatchSnapshot()
    })

    test('in filter with 3 arg new notation', () => {
      const { sql, values } = cqn2sql({
        SELECT: {
          from: { ref: ['Foo'] },
          where: [{ ref: ['c'] }, '=', { func: 'concat', args: [{ val: 'Existing' }, { ref: ['a'] }, { val: '!' }] }],
        },
      })
      expect({ sql, values }).toMatchSnapshot()
    })

    test('in filter with nested functions new notation', () => {
      const { sql, values } = cqn2sql({
        SELECT: {
          from: { ref: ['Foo'] },
          where: [
            { func: 'lower', args: [{ ref: ['a'] }] },
            '=',
            {
              func: 'lower',
              args: [{ func: 'upper', args: [{ func: 'trim', args: [{ val: '   existing name  ' }] }] }],
            },
            'and',
            { func: 'length', args: [{ func: 'trim', args: [{ val: '  name' }] }] },
            '=',
            { ref: ['b'] },
          ],
        },
      })
      expect({ sql, values }).toMatchSnapshot()
    })

    test('in filter with subselect as function param', () => {
      const subselect = {
        SELECT: {
          columns: [{ ref: ['ID'] }],
          from: { ref: ['Foo2'] },
          where: [{ ref: ['a'] }, '=', { val: 1 }],
        },
      }
      const { sql } = cqn2sql({
        SELECT: {
          from: { ref: ['Foo'] },
          where: [{ ref: ['ID'] }, '=', { func: 'any', args: [subselect] }],
        },
      })
      expect(sql).toMatchSnapshot()
    })
  })

  describe('quoted column aliases', () => {
    // aliases should be quoted only for HANA
    test('simple select with column aliases', () => {
      const cqn = {
        SELECT: {
          from: { ref: ['Foo'], as: 'T' },
          columns: [
            { ref: ['a'], as: 'A' },
            { val: true, as: 'True' },
            { val: false, as: 'False' },
            { val: null, as: 'Null' },
            { func: 'count', args: ['*'], as: 'CountFunc' },
          ],
        },
      }
      const { sql } = cqn2sql(cqn)
      expect({ sql }).toMatchSnapshot()
    })

    // aliases should be quoted only for HANA
    test('select with subselect with in and column aliases', () => {
      const { sql, values } = cqn2sql(cqn.aliasWithInSubSelect)
      expect({ sql, values }).toMatchSnapshot()
    })

    // aliases should be quoted only for HANA
    test('select with subselect in exists and column aliases', () => {
      const { sql } = cqn2sql(cqn.aliasWithNestedExists)
      expect(sql).toMatchSnapshot()
    })

    // aliases should be quoted only for HANA
    test('select with simple subselect and column aliases', () => {
      const { sql } = cqn2sql(cqn.aliasWithSubSelect)
      expect(sql).toMatchSnapshot()
    })
  })

  describe('joins', () => {
    test.skip('with table join table', () => {
      const { sql } = cqn2sql(cqn.join)
      expect(sql).toMatchSnapshot()
    })
  })
})
