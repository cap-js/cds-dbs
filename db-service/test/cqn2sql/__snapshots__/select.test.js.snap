// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cqn2sql GROUP BY GROUP BY two columns 1`] = `"SELECT Foo.a,Foo.b FROM Foo as Foo GROUP BY Foo.x,Foo.c"`;

exports[`cqn2sql HAVING clauses with select specific elements with from type string with having clause 1`] = `"SELECT Foo.a,Foo.b,Foo.c FROM Foo as Foo HAVING Foo.x < 9"`;

exports[`cqn2sql LIMIT with limit and offset 1`] = `"SELECT Foo.a,Foo.b,Foo.c FROM Foo as Foo LIMIT 1 OFFSET 2"`;

exports[`cqn2sql LIMIT with limit without offset 1`] = `"SELECT Foo.a,Foo.b,Foo.c FROM Foo as Foo LIMIT 1"`;

exports[`cqn2sql ONE one results in limit 1 1`] = `"SELECT Foo.a,Foo.b,Foo.c FROM Foo as Foo LIMIT 1"`;

exports[`cqn2sql ORDER BY ORDER BY alias 1`] = `"SELECT Foo.a,Foo.b,count(Foo.x) as count1 FROM Foo as Foo ORDER BY count1 ASC"`;

exports[`cqn2sql WHERE EXISTS with nested EXISTS 1`] = `"SELECT T2.ID,T2.a,T2.b,T2.c,T2.x FROM Foo as T2 WHERE exists (SELECT 1 FROM Books as T1 WHERE T1.ID = 1 and exists (SELECT 1 FROM Foo2 as T0 WHERE T0.ID = 11 and T1.ID = T0.a) and T2.ID = T1.ID)"`;

exports[`cqn2sql WHERE entries where one column holds entries smaller than 9 1`] = `"SELECT Foo.ID,Foo.a,Foo.b,Foo.c,Foo.x FROM Foo as Foo WHERE Foo.x < 9"`;

exports[`cqn2sql WHERE entries where one column holds entries which are in list 1`] = `"SELECT Foo.ID,Foo.a,Foo.b,Foo.c,Foo.x FROM Foo as Foo WHERE Foo.x IN (1,2,3)"`;

exports[`cqn2sql WHERE entries where with int reference and param true 1`] = `
{
  "sql": "SELECT Foo.ID,Foo.a,Foo.b,Foo.c,Foo.x FROM Foo as Foo WHERE Foo.x = :7",
  "values": [],
}
`;

exports[`cqn2sql WHERE entries where with place holder 1`] = `
{
  "sql": "SELECT Foo.ID,Foo.a,Foo.b,Foo.c,Foo.x FROM Foo as Foo WHERE Foo.ID = ?",
  "values": [],
}
`;

exports[`cqn2sql WHERE select with a nested select in a complex where 1`] = `"SELECT Foo.a,Foo.b,Foo.c FROM Foo as Foo WHERE ( Foo.x + 1 ) < 9 AND Foo.x IN (SELECT Foo.a FROM Foo as Foo WHERE Foo.x < 9)"`;

exports[`cqn2sql WHERE select with a nested select in where 1`] = `"SELECT Foo.a,Foo.b,Foo.c FROM Foo as Foo WHERE Foo.x IN (SELECT Foo.a FROM Foo as Foo WHERE Foo.x < 9)"`;

exports[`cqn2sql WHERE where with partial cqn 1`] = `"SELECT Foo.ID,Foo.a,Foo.b,Foo.c,Foo.x FROM Foo as Foo WHERE (Foo.x = 9)"`;

exports[`cqn2sql WHERE where with two partial cqn 1`] = `"SELECT Foo.ID,Foo.a,Foo.b,Foo.c,Foo.x FROM Foo as Foo WHERE (Foo.x + 9) = 9"`;

exports[`cqn2sql WHERE with contains with multiple arguments 1`] = `"SELECT Foo.ID,Foo.a,Foo.b,Foo.c,Foo.x FROM Foo as Foo WHERE Foo.a = 0 and ifnull(instr((Foo.a,Foo.b,Foo.c,Foo.x),?),0)"`;

exports[`cqn2sql WHERE with contains with one column in where clause 1`] = `
{
  "sql": "SELECT Foo.ID,Foo.a,Foo.b,Foo.c,Foo.x FROM Foo as Foo WHERE ifnull(instr((Foo.b),?),0)",
  "values": [
    "5",
  ],
}
`;

exports[`cqn2sql WHERE with list of values 1`] = `
{
  "sql": "SELECT Foo.ID,Foo.a,Foo.b,Foo.c,Foo.x FROM Foo as Foo WHERE (Foo.a,Foo.b,1) = (Foo.c,?,Foo.x)",
  "values": [
    "d",
  ],
}
`;

exports[`cqn2sql WHERE with select with exist in where condition 1`] = `"SELECT T1.ID,T1.a,T1.b,T1.c,T1.x FROM Foo as T1 WHERE exists (SELECT Foo2.ID,Foo2.name,Foo2.a FROM Foo2 as Foo2)"`;

exports[`cqn2sql aggregation functions with select with count(1) 1`] = `"SELECT count(1) as count FROM Foo as Foo"`;

exports[`cqn2sql aggregation functions with select with different functions without alias in elements 1`] = `"SELECT min(Foo.x) as min,count(1) as count,sum(Foo.x) as sum FROM Foo as Foo"`;

exports[`cqn2sql aggregation functions with select with functions in elements new notation 1`] = `"SELECT min(Foo.x) as foo1,Foo.a,count(*) as foo2,count(1) as foo3,sum(Foo.x) as foo4 FROM Foo as Foo"`;

exports[`cqn2sql aggregation functions with select with functions in where clause new notation 1`] = `"SELECT Foo.a,Foo.b,Foo.c FROM Foo as Foo WHERE max(Foo.x) < 9"`;

exports[`cqn2sql complex combinations AS, sub query 1`] = `"SELECT Foo.a,Foo.b as B,1 as C,Foo.x + 2 as D,(SELECT Foo2.ID,Foo2.a,Foo2.b,Foo2.c,Foo2.x FROM Foo as Foo2) as E FROM Foo as Foo"`;

exports[`cqn2sql complex combinations Exists in object mode in complex where 1`] = `
{
  "sql": "SELECT Foo.ID,Foo.a,Foo.b,Foo.c,Foo.x FROM Foo as Foo WHERE Foo.ID = ? and ( exists (SELECT Author.id FROM Author as Author WHERE Author.id <> ?) or exists (SELECT Foo2.ID FROM Foo2 as Foo2 WHERE Foo2.name is distinct from ?) )",
  "values": [
    "123",
    "",
    "",
  ],
}
`;

exports[`cqn2sql complex combinations WHERE, GROUP BY, HAVING, ORDER BY, LIMIT, OFFSET 1`] = `"SELECT Foo.x + 1 as foo1,Foo.b,Foo.c FROM Foo as Foo WHERE Foo.ID = 111 GROUP BY Foo.x HAVING Foo.x < 9 ORDER BY c ASC LIMIT 11 OFFSET 22"`;

exports[`cqn2sql functions new notation in filter with 1 arg new notation 1`] = `
{
  "sql": "SELECT Foo.ID,Foo.a,Foo.b,Foo.c,Foo.x FROM Foo as Foo WHERE lower(Foo.c) = ?",
  "values": [
    "name",
  ],
}
`;

exports[`cqn2sql functions new notation in filter with 2 arg new notation 1`] = `"SELECT Foo.ID,Foo.a,Foo.b,Foo.c,Foo.x FROM Foo as Foo WHERE Foo.c = Foo.a || Foo.b"`;

exports[`cqn2sql functions new notation in filter with 3 arg new notation 1`] = `
{
  "sql": "SELECT Foo.ID,Foo.a,Foo.b,Foo.c,Foo.x FROM Foo as Foo WHERE Foo.c = ? || Foo.a || ?",
  "values": [
    "Existing",
    "!",
  ],
}
`;

exports[`cqn2sql functions new notation in filter with asterisk as arg new notation 1`] = `"SELECT Foo.ID,Foo.a,Foo.b,Foo.c,Foo.x FROM Foo as Foo HAVING count(*) > 1"`;

exports[`cqn2sql functions new notation in filter with nested functions new notation 1`] = `
{
  "sql": "SELECT Foo.ID,Foo.a,Foo.b,Foo.c,Foo.x FROM Foo as Foo WHERE lower(Foo.a) = lower(upper(trim(?))) and length(trim(?)) = Foo.b",
  "values": [
    "   existing name  ",
    "  name",
  ],
}
`;

exports[`cqn2sql functions new notation in filter with subselect as function param 1`] = `"SELECT Foo.ID,Foo.a,Foo.b,Foo.c,Foo.x FROM Foo as Foo WHERE Foo.ID = any((SELECT Foo2.ID FROM Foo2 as Foo2 WHERE Foo2.a = 1))"`;

exports[`cqn2sql functions new notation in orderby with 1 arg new notation 1`] = `"SELECT Foo.ID,Foo.a,Foo.b,Foo.c,Foo.x FROM Foo as Foo ORDER BY lower(Foo.c) DESC"`;

exports[`cqn2sql quoted column aliases select with simple subselect and column aliases 1`] = `"SELECT Foo.a,Foo.b as B,1 as C,Foo.x + 2 as D,(SELECT Foo2.a,Foo2.b as B,false as False,Foo2.x + 2 as Xpr FROM Foo as Foo2) as E FROM Foo as Foo"`;

exports[`cqn2sql quoted column aliases select with subselect in exists and column aliases 1`] = `"SELECT T2.id,T2.version,T2.parent_ID FROM Author as T2 WHERE exists (SELECT 1 as One,T1.code as Xpr1 FROM Books as T1 WHERE T1.ID = 1 and exists (SELECT 3 as Three,T0.x + 1 as Xpr2 FROM Foo as T0 WHERE T0.ID = 11 and T1.ID = T0.b))"`;

exports[`cqn2sql quoted column aliases select with subselect with in and column aliases 1`] = `
{
  "sql": "SELECT Foo.a as A,? as ABC,Foo.x + 1 as Xpr1 FROM Foo as Foo WHERE ( Foo.x + 1 ) < 9 AND Foo.x IN (SELECT Foo.a as B,Foo.x - 4 as Xpr2 FROM Foo as Foo WHERE Foo.x < 9)",
  "values": [
    "abc",
  ],
}
`;

exports[`cqn2sql quoted column aliases simple select with column aliases 1`] = `
{
  "sql": "SELECT T.a as A,true as True,false as False,NULL as Null,count(*) as CountFunc FROM Foo as T",
}
`;

exports[`cqn2sql selection of columns of one table select distinct 1`] = `"SELECT DISTINCT Foo.a,Foo.b,Foo.c FROM Foo as Foo"`;

exports[`cqn2sql selection of columns of one table with select from non existent entity with star wildcard 1`] = `"SELECT * FROM "Â¿HoWdIdYoUmAnAgeToCaLaNeNtItyThIsNaMe?""`;

exports[`cqn2sql selection of columns of one table with select specific elements with from ref 1`] = `"SELECT Foo.a,Foo.b,Foo.c FROM Foo as Foo"`;

exports[`cqn2sql selection of columns of one table with select that has (faked) reflection model 1`] = `"SELECT T1.ID,T1.a,T1.b,T1.c,T1.x FROM Foo as T1"`;

exports[`cqn2sql selection of columns of one table with select with asterisk in columns 1`] = `"SELECT Foo.ID,Foo.a,Foo.b,Foo.c,Foo.x FROM Foo as Foo"`;

exports[`cqn2sql selection of columns of one table with select with empty orderBy and specific elements with from type string 1`] = `"SELECT Foo.a,Foo.b,Foo.c FROM Foo as Foo"`;

exports[`cqn2sql selection of columns of one table with select with from ref and elements = undefined 1`] = `"SELECT Foo.ID,Foo.a,Foo.b,Foo.c,Foo.x FROM Foo as Foo"`;

exports[`cqn2sql selection of columns of one table with select with from ref and elements as empty array 1`] = `"SELECT FROM Foo as Foo"`;
